json editor case can not compile

I encountered some problem when I tried to run "Json Editor" case.

The code segment is:
```rust
fn run_app<B: Backend>(terminal: &mut Terminal<B>, app: &mut App) -> io::Result<bool>
{
    loop {
        terminal.draw(|f| ui::ui(f, app))?;
        // ...
    }
}
```

Problem is `Terminal::draw()` returns a result with type `Result<CompletedFrame<'_>, <B as Backend>::Error>`, but `run_app` function returns `io::Result<bool>` which is not compatible with type returned by `Terminal::draw`.

I wondered how could I fix this?

I tried this but it didn't work:
```rust
fn run_app<B: Backend>(terminal: &mut Terminal<B>, app: &mut App) -> io::Result<bool>
where <B as Backend>::Error: Into<std::io::Error>
{
    loop {
        terminal.draw(|f| ui::ui(f, app))?;
        // ...
    }
}
```

Compile error showed as below:
```bash
error[E0277]: `?` couldn't convert the error to `std::io::Error`
  --> src-json-editor\main.rs:52:42
   |
48 | fn run_app<B: Backend>(terminal: &mut Terminal<B>, app: &mut App) -> io::Result<bool>
   |                                                                      ---------------- expected `std::io::Error` because of this
...
52 |         terminal.draw(|f| ui::ui(f, app))?;
   |                  ------------------------^ the trait `From<<B as Backend>::Error>` is not implemented for `std::io::Error`
   |                  |
   |                  this can't be annotated with `?` because it has type `Result<_, <B as Backend>::Error>`
   |
   = note: the question mark operation (`?`) implicitly performs a conversion on the error value using the `From` trait
help: consider extending the `where` clause, but there might be an alternative better way to express this requirement
   |
49 | where <B as Backend>::Error : Into<std::io::Error>, std::io::Error: From<<B as Backend>::Error>
   |                                                   +++++++++++++++++++++++++++++++++++++++++++++
```

--- a/code/tutorials/json-editor/Cargo.toml
+++ b/code/tutorials/json-editor/Cargo.toml
@@ -9,6 +9,6 @@ publish.workspace = true
 # See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
 
 [dependencies]
-ratatui = "0.29.0"
+ratatui = "0.30.0"
 serde = { version = "1.0.219", features = ["derive"] }
 serde_json = "1.0.140"

--- a/code/tutorials/json-editor/src/main.rs
+++ b/code/tutorials/json-editor/src/main.rs
@@ -62,7 +62,10 @@ fn main() -> Result<(), Box<dyn Error>> {
 
 // ANCHOR: run_app_all
 // ANCHOR: run_method_signature
-fn run_app<B: Backend>(terminal: &mut Terminal<B>, app: &mut App) -> io::Result<bool> {
+fn run_app<B: Backend>(terminal: &mut Terminal<B>, app: &mut App) -> io::Result<bool>
+where
+    io::Error: From<B::Error>,
+{
     // ANCHOR_END: run_method_signature
     // ANCHOR: ui_loop
     loop {

--- a/src/content/docs/tutorials/json-editor/main.md
+++ b/src/content/docs/tutorials/json-editor/main.md
@@ -119,9 +119,13 @@ included in `ratatui`. It allows for 3rd party backends to be implemented.
 
 `run_app` also requires a mutable borrow to an application state object, as defined in this project.
 
-Finally, the `run_app` returns an `io::Result<bool>` that indicates if there was an io error with
-the `Err` state, and an `Ok(true)` or `Ok(false)` that indicates if the program should print out the
-finished json.
+The return type is still `io::Result<bool>`, but there is one extra bound:
+`io::Error: From<B::Error>`. This is needed because `Terminal::draw()` returns a `Result` using the
+backend's error type, and `?` can only propagate that error if it can be converted into `io::Error`.
+
+With that conversion in place, `run_app` can use `?` on both `event::read()` and
+`terminal.draw(...)`, while still returning `Ok(true)` or `Ok(false)` to indicate whether the
+finished json should be printed.
 
 ### UI Loop
 

